src/D3MHub.sol

```diff
@@ -395,126 +390,44 @@ contract D3MHub {
         @param ilk bytes32 of the D3M ilk name
     */
     function exec(bytes32 ilk) external lock {
+        // IMPORTANT: this function assumes Vat rate of D3M ilks will always be == 1 * RAY (no fees).
+        // That's why this module converts normalized debt (art) to Vat DAI generated with a simple RAY multiplication or division
+
         (uint256 Art, uint256 rate, uint256 spot, uint256 line,) = vat.ilks(ilk);
         require(rate == RAY, "D3MHub/rate-not-one");
         require(spot == RAY, "D3MHub/spot-not-one");
 
         ID3MPool _pool = ilks[ilk].pool;
 
-        _pool.preDebtChange("exec");
-        uint256 currentAssets = _pool.assetBalance();
+        _pool.preDebtChange();
 
         if (vat.live() == 0) {
             // MCD caged
-            require(end.debt() == 0, "D3MHub/end-debt-already-set");
+            EndLike _end = end;
+            require(_end.debt() == 0, "D3MHub/end-debt-already-set");
```

Comment:
We should document the consequence for the system, either here in comments or the README, if we fail to call `exec` before `End.debt` gets set.

IIUC, that consequence is a haircut to DAI holders in terms of disbursed collateral based on the fraction of the total debt the D3M represents. Although I think the actual market dynamics can be much less straightforward, since only D3M DAI that has been borrowed in the lending market is "real" and borrowers have an incentive to pay up base on AAVE collateralization to obtain DAI and unwind. Take the example of the only collateral is a D3M with N DAI deposited in AAVE, one borrower who has borrowed M < N DAI against ETH, and then sold that M DAI to someone else who is holding it. If the system is caged and `End.thaw()` is called without a successful `skim` of the D3M urn...strictly speaking, the DAI holder gets nothing from the system. BUT, if the AAVE borrower wants his full collateral back, and AAVE's min health factor is H > 1, he's willing to pay $H / DAI to the DAI holder, so the DAI holder actually is more than made whole. In fact if AAVE has a live DAI oracle...the price going up affects the poor borrower's collateralization ratio, so he might have to pay more than H or even be liquidated...

Going further: neither `End.skim()` nor `_unwind()` actually changes `Vat.debt` AFAICT, so I am not convinced we even need this check here! Unless `Pool.postDebtChange` might modify `Vat.debt`...

Reply:
So there are two ways to redeem this after shutdown:

1) Just stop `exec` on shutdown and share `aDai` as any other collateral to DAI holders. We could in theory do this, it is the simplest and equals to any other collateral. It is not great though as requires infinite cyclical DAI redemption as we have for example with DAI LPs.

2) Try to recover the most DAI we can, this way we reduce `aDai` shares and the DAI recovered is going to surplus, which is burnt after shutdown, reducing `vat.debt`, that's why we can only do this process until `end.thaw()` is executed.

I just glanced the example but I think I replied the main question, if it is not the case will go deeply on it.

Reply:
I think I answered my own question about the need to do this...since we transfer dai to the Vow, someone could then call `heal` and change `debt`.

IIUC, case 1. can still happen even if we fail to execute this flow? Think I missed that in my analysis.

Reply:
Yeah case 1 can still happen if nobody executes `exec` during the `end.wait` time.
Anyone can do `end.skim(ilk)`, that doesn't change `vat.debt`, just `vat.vice` and `vat.sin(vow)` (which are not used in any `End` distribution accounting). The only result should be that now the `end` has `aDai` collateral to share the `DAI` holders.

The advantage of calling `exec` before `end.thaw` is actually trying to unwind collateral before settlement, in that case it will in fact actually affect `vat.debt` for collateral distribution. There will be less DAI in circulation (`vat.debt`) but also less (or none) `aDai` to give to the DAI holders.
In case we do not call it we will have more `aDai` with more DAI in circulation, which in fact if you cycle in an infinite way from DAI -> aDai -> DAI -> ... you should end up with the same amount of the other collaterals in both cases (assuming there is liquidity to withdraw DAI from aDai).